### PR TITLE
Type test improvements

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Make a few changes to type test generation:

- Internal refactoring to compute things like the test name once
- Add type test for class statics
- Remove redundant "Declaration" from name of all type tests
- Treat `typeof` cases like sealed: removing checks for removal of members of variables, functions and class statics.

This means that going from:

```typescript
function foo(): 1 | 2 { return 1;}
const thing = {a: "a"} as const;
```
To
```typescript
function foo(optional?: string): 1 { return 1;}
const thing = {a: "a", b: "b" } as const;
```

Used to be breaking, but now should not be.
Making the types of variables/functions less specific is still a breaking change (they are still type tested, just treated like sealed types).

Note that:

```typescrtipt
type Foo = () => 1|2;
```
To
```typescrtipt
type Foo = () => 1;
```

Is still breaking: that's a type definition not the type of a function definition, so no changes are made to the policy there.

Changes for client can be viewed here: https://github.com/microsoft/FluidFramework/pull/21875

## Breaking Changes

Now catches changes to class statics, and broken annotation in package.json need to have the "Declaration" removed from he names.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

### CODE REVIEW

File: build-tools/packages/build-tools/src/typeValidator/typeData.ts [hunk 1]
```diff
@@ -3,23 +3,26 @@
...
 	readonly tags: ReadonlySet<string>;
-}
-
-/**
- * Creates a non-colliding "mangled" identifier for the type.
- */
-export function getFullTypeName(typeData: TypeData) {
-	return `${typeData.kind}_${typeData.name.replace(/\./g, "_")}`;
+	/**
+	 * Does this TypeData to the the name item, or the typeof the name item.
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/typeValidator/typeData.ts:22]: I think you're missing a verb here but I don't know which one. "belong"?

File: build-tools/packages/build-cli/src/commands/generate/typetests.ts [hunk 2]
```diff
@@ -357,7 +356,7 @@ export function typeDataFromFile(
 				log,
 				addStringScope(namespacePrefix, exportedName),
 			)) {
-				const fullName = getFullTypeName(typeDefinition);
+				const fullName = typeDefinition.testCaseName;
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/generate/typetests.ts:359]: nice

**Review by tylerbutler — APPROVED** [review]
Typo in one of the docstrings, and I didn't look super-closely at the logic details, instead trusting the PR with the diff on main, which was super helpful.

File: build-tools/packages/build-tools/src/typeValidator/typeData.ts (hunk 1 above)
Comment (CraigMacomber) [reply build-tools/packages/build-tools/src/typeValidator/typeData.ts:22]:
```suggestion
	 * Indicates if this TypeData refer to the named item (false), or the typeof the named item (true).
```

**Pull request merged**